### ISSUE TITLE
Parallel quasi-newton optimizers for tensorflow-gpu

### ISSUE THREAD

**Issue by timudk** [tensorflow/tensorflow#26100]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 1.12.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Is there any interest in implementing quasi-Newton methods for tensorflow-gpu? There has been some work on a parallel L-BFGS-B method (https://www.sciencedirect.com/science/article/pii/S0097849314000119) for GPus.

**Will this change the current api? How?**
New implementation.

**Who will benefit with this feature?**
I have had some recent success with using the BFGS optimizer for feedforward neural network. This approach, however, is limited to small networks and to tensorflow-cpu.

**Any Other info.**

**Comment by timudk** [comment]
Any updates?

**Comment by JiaweiZhuang** [comment]
Found this from a [tweet](https://twitter.com/timudk/status/1118240779170803712). Second-order methods are very efficient for small & smooth problems. An example is [Picking an Optimizer for Style Transfer](https://blog.slavv.com/picking-an-optimizer-for-style-transfer-86e7b8cba84b) that shows BFGS performs better than SGD-like methods (related: tensorflow/tfjs#127). Currently this is done through a SciPy wrapper without GPU support (keras-team/keras##5085)

Useful references: 
- PyTorch's L-BFGS implementation : [source code](https://pytorch.org/docs/stable/_modules/torch/optim/lbfgs.html) and [usage in style transfer](https://pytorch.org/tutorials/advanced/neural_style_tutorial.html#gradient-descent).
- Prototyping L-BFGS with TF-eager: [TensorFlow meets PyTorch with Eager execution](https://medium.com/@yaroslavvb/tensorflow-meets-pytorch-with-eager-mode-714cce161e6c).

**Comment by junpenglao** [comment]
You can have a look at the implementation of BFGS and LBFGS in Tensorflow_probability, which do not wrap Scipy: https://github.com/tensorflow/probability/tree/master/tensorflow_probability/python/optimizer

**Comment by junpenglao** [comment]
It might be better to raise this feature request to Tensorflow_probability.

**Comment by timudk** [comment]
I finally got TFP's BFGS optimizer working for a supervised problem. Would adding an example to https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples be desirable?

**Comment by sanjoy** [comment]
> I finally got TFP's BFGS optimizer working for a supervised problem. Would adding an example to https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples be desirable?

@jvdillon WDYT?

**Comment by danieltsai0** [comment]
> I finally got TFP's BFGS optimizer working for a supervised problem. Would adding an example to https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples be desirable?

Can you share a code-snippet?

**Comment by rmothukuru** [comment]
@timudk,
As per [this comment](https://github.com/tensorflow/tensorflow/issues/26100#issuecomment-505904074) this Feature seems to be implemented. Can you please confirm the same so that we can close this issue? Thanks!
